build/src/lib.rs

```diff
@@ -1,26 +1,62 @@
 //! This crate provides a builder to make it easier to use the correct binaries for [nlprule](https://github.com/bminixhofer/nlprule).
 //! See `README.md` for details.
 
-use flate2::read::GzDecoder;
+use flate2::bufread::GzDecoder;
+use fs::File;
+use fs_err as fs;
 use nlprule::{compile, rules_filename, tokenizer_filename};
+use std::fs::Permissions;
 use std::{
-    fs::{self, File},
-    io::{self, BufWriter, Cursor, Read},
+    io::{self, BufReader, BufWriter, Cursor, Read, Seek, SeekFrom},
     path::{Path, PathBuf},
+    result,
 };
-use thiserror::Error;
 use zip::result::ZipError;
 
-#[derive(Debug, Error)]
+#[derive(Debug, thiserror::Error)]
 pub enum Error {
-    #[error("request error: {0}")]
+    #[error(transparent)]
     RequestError(#[from] reqwest::Error),
-    #[error("i/o error: {0}")]
+    #[error("Binaries were not found on the remote")]
+    BinariesNotFound,
+    #[error("Failed to validate {1:?} binary for lang {0}")]
+    ValidationFailed(String, Binary, #[source] nlprule::Error),
+    #[error(transparent)]
     IOError(#[from] io::Error),
-    #[error("zip error: {0}")]
+    #[error(transparent)]
     ZipError(#[from] ZipError),
     #[error("error postprocessing binaries: {0}")]
-    PostprocessingError(Box<dyn std::error::Error>),
+    PostprocessingError(#[from] Box<dyn std::error::Error + Send + Sync + 'static>),
+}
+
+pub type Result<T> = std::result::Result<T, Error>;
+
+/// Definition of the data transformation for the network retrieved, binencoded rules and tokenizer datasets.
+pub trait TransformDataFn<I: Read>:
```

Comment:
That's what I meant in my previous comment in the PR (not in the review pane). That's complexity to be removed.

I don't think it makes sense to use type aliases. Passing the type would then require the user to Box or have the trait bounds there (and hence twice).

While in this case one could argue, that a simplified `fn` signature might be good, I find that a bit of an anti-pattern. `impl Read` and `impl Write` can easily be chained without intermediate buffers.

Reply:
I just attempted it, but having an `dyn Fn(dyn Read, dyn Write) -> result::Result<(), OtherError> {}` opens a can of wurms, so I am not sure that's in scope of this PR.

Reply:
You're right, I agree on the `impl Read` / `impl Write` now, it is also consistent with the signature of `postprocess`.

> Passing the type would then require the user to Box or have the trait bounds there (and hence twice).

I don't see an issue with having the bounds twice (once in the type alias, once in the `transform` function). Something like this would work:

```rust
pub type TransformDataFn =
    Box<dyn Fn(Cursor<Vec<u8>>, Cursor<&mut Vec<u8>>) -> result::Result<(), OtherError>>;
```

and 

```rust
pub fn transform<F, C>(mut self, proc_fn: C, path_fn: F) -> Self
where
    C: Fn(Cursor<Vec<u8>>, Cursor<&mut Vec<u8>>) -> result::Result<(), OtherError> + 'static,
    F: TransformPathFn + 'static,
{
    self.transform_data_fn = Some(Box::new(proc_fn));
    self.transform_path_fn = Some(Box::new(path_fn));
    self
}
```

or am I missing something? The repetition is not perfect but changing one or the other immediately gives a compile time error so that's not an issue.

Reply:
See the last commit, one needs late bound life times and a bunch of boxes to be able to use `dyn Read` and `dyn Write` in the args which is needed for the `get_build_dir` case.